src/inmanta/execute/runtime.py

```diff
@@ -69,15 +69,33 @@ def receive_result(self, value: T, location: Location) -> None:
         raise NotImplementedError()
 
 
-class IPromise(Generic[T]):
+class IPromise:
+    """
+    A promise to the owner to provide a value or progression towards a value in some way, either directly or indirectly.
+    To provide strict provider tracking, overpromising is allowed in case of uncertainty as to whether progression towards
+    a value will be made, as long as progression towards certainty is made.
+    """
+
+    __slots__ = ()
+
+
+class ISetPromise(IPromise, Generic[T]):
+    """
+    A promise to the owner to set a value.
+    """
+
     __slots__ = ()
 
+    # TODO: can recur be removed? If not, document! Might be required only for list promise
```

Comment:
recur is for relations, to set the forward relation, then reverse relation and then stop.